Boost Update pull request

psbritt opened the pull request:
All changes to allow for Boost 1.72, removing the requirement for usage of the old ABI when building FRENSIE, have been made.

aprobinson commented on CMakeLists.txt:
```diff
@@ -1,6 +1,9 @@
 ## Initialize the FRENSIE Project
 ##---------------------------------------------------------------------------##
+# Define the project
+project(FRENSIE)
```
Why was this line added? The included file below (SetProjectName) calls this.

psbritt commented:
I may have forgotten to mention that I updated to the most recent version of CMake. When I originally compiled with the newest version, it gave a warning about not having a project name before whatever the first line originally was, so I added that line which fixed the warning. I’ll investigate a bit more tomorrow to see what the best way to fix it such that the warning doesn’t come back and there isn’t a redundant code.



> On Apr 5, 2020, at 9:44 PM, Alex Robinson <notifications@github.com> wrote:
> 
> ﻿
> @aprobinson requested changes on this pull request.
> 
> I think the added project(FRENSIE) at the beginning of the top CMakeLists.txt file is redundant. Let me know if you disagree (also see my comment directly in the file for the PR). If not please remove before merging. Everything else looks good to me. I'll attempt a build but I'm assuming that all of the tests are passing for you.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

aprobinson commented:
Got it. If that's the case then I would recommend taking whats in ProjectName.cmake (cmake directory) and moving it directly to the beginning of the CMakeLists.txt file. Then you can remove ProjectName.cmake.

aprobinson commented:
Also, just realized. There is a scripts directory with some shell scripts that can be used for building. Please remove the ABI lines from those scripts before merging.

aprobinson commented:
What version of moab are you building with? I'm seeing the following error in the utility_mesh package:

In file included from /home/alexr/Research/transport/forks/psbritt/src/packages/utility/mesh/src/Utility_TetMesh.cpp:28:0:
/home/alexr/Software/moab/include/moab/Matrix3.hpp:51:10: fatal error: Eigen/Dense: No such file or directory
 #include "Eigen/Dense"
          ^~~~~~~~~~~~~
compilation terminated.

psbritt commented:
I'm using version 5.1.0

psbritt commented:
> Got it. If that's the case then I would recommend taking whats in ProjectName.cmake (cmake directory) and moving it directly to the beginning of the CMakeLists.txt file. Then you can remove ProjectName.cmake.

Unfortunately this solution introduces a cmake error. I'll investigate.

psbritt commented:
I'm getting weird compiler errors now too, but not that one. I'm investigating now.

psbritt commented:
Basically what I'm getting right now is that the cmake function COMPARE_COMPILERS is getting incorrect arguments. I'm noticing that the variable CMAKE_C_COMPILER is never set, so when I went to figure out what it was, there was no value, while MOAB was compiled with gcc. This variable is set in the enable_MPI_support macro, but why is it not set in the enable_moab_support macro?

psbritt commented:
The cmake change recommended breaks the link between FRENSIE and MOAB. When attempting to run frensie.sh, it states that a compiler_comparison function is passed the wrong arguments. Something is important about the order of these CMake commands - I'm just not sure what.

psbritt commented:
The project command has to be implemented in the main CMakeLists.txt file for new versions, and an INCLUDE statement with a file that contains it is no longer sufficient. My addition of PROJECT(FRENSIE) included the compilers for C++ and C, which are default compilers for PROJECT(). The other command only included C++ and Fortran. I simply added C to the original PROJECT() command and moved it to the main CMakeLists.txt file to fix it.

Sidenote: Why are we checking the MOAB C compiler to make sure it's the same as the C compiler used by CMake for FRENSIE when, to my knowledge, we don't use C at all in FRENSIE?

lkersting commented:
The `project()` CMake function initializes a lot of other CMake functions under the hood. If no language/compiler is specified I believe it defaults to C and C++. It is important to note that once a compiler is set it shouldn't/cannot be changed in the CMakeLists.txt file. A language can be enabled after the `project()` function is called by using the CMake command `enable_language()` (i.e. `enable_language(CXX)`. This might be important for MPI builds if it requires a different compiler.

psbritt commented:
I understand all of that, but this isn't making it clear to me why we're checking the equivalence of the C compilers between MOAB and FRENSIE when FRENSIE doesn't use C.

aprobinson commented:
There was a reason Philip. If I remember correctly it had to do with MPI, which does use the C compilers as it's a C library (at least we used to use the C interfaces like most people). Now that FRENSIE uses boost mpi it's not as relevant. I think that it's included as a sanity check to make sure that the same compilers were used to build everything. Otherwise there's the potential for errors at link time.

gonuke commented:
FWIW- I'd recommend that you require a different person to merge a PR (e.g. @aprobinson ) that the person who proposed it (@psbritt ).

aprobinson commented:
Thanks Paul (@gonuke), that's fair

lkersting commented:
I believe it is also possible to require approval for merge requests.
